### PR TITLE
Lock around usages of imaging memory arenas

### PR DESCRIPTION
**Pull request by lysnikolaou**
Related to #8199.

Did some testing with threads. Thread-sanitizer is also happy after these changes.

### CODE REVIEW

File: src/libImaging/Storage.c [hunk 1]
```diff
@@ -365,17 +368,18 @@ ImagingDestroyArray(Imaging im) {
 
     if (im->blocks) {
         while (im->blocks[y].ptr) {
+            MUTEX_LOCK(&ImagingDefaultArena.mutex);
             memory_return_block(&ImagingDefaultArena, im->blocks[y]);
+            MUTEX_UNLOCK(&ImagingDefaultArena.mutex);
             y += 1;
         }
```
Comment (Yay295) [inline src/libImaging/Storage.c:377]: It would probably be better to move the lock/unlock outside the loop.

File: src/_imaging.c [hunk 2]
```diff
@@ -4071,14 +4091,20 @@ _set_blocks_max(PyObject *self, PyObject *args) {
     if (blocks_max < 0) {
         PyErr_SetString(PyExc_ValueError, "blocks_max should be greater than 0");
         return NULL;
-    } else if (
-        (unsigned long)blocks_max >
-        SIZE_MAX / sizeof(ImagingDefaultArena.blocks_pool[0])) {
+    }
+
+    MUTEX_LOCK(&ImagingDefaultArena.mutex);
+    size_t blocksize = sizeof(ImagingDefaultArena.blocks_pool[0]);
```
Comment (nulano) [inline src/_imaging.c:4097]:
```c
sizeof(ImagingDefaultArena.blocks_pool[0])
````
is evaluated at compile time so there is no need to lock here.

In fact, you could even use
```c
sizeof(((ImagingMemoryArena*)0)->blocks_pool[0])
```

Comment (lysnikolaou) [reply src/_imaging.c]: Thanks for the pointer! Updated.

**Comment by lysnikolaou** [comment]
This is ready to be reviewed more throughly. All tests pass and tsan is happy.

**Comment by nulano** [comment]
> All tests pass and tsan is happy.

Out of curiosity, how are you testing with tsan? I don't think we have it in CI (although perhaps we should add it!).

**Comment by lysnikolaou** [comment]
> Out of curiosity, how are you testing with tsan? I don't think we have it in CI (although perhaps we should add it!).

I added a few very very basic threaded tests, built CPython and Pillow with tsan enabled and then ran the tests. It all takes forever, cause CPython with tsan is very slow, but it works.

If your question is how I build Pillow with tsan enabled, it is really easy:
```
CC="clang -fsanitize=thread" python -m pip install .
```

**Review by hugovk — APPROVED** [review]

**Comment by hugovk** [comment]
If no other reviews or comments, I'll merge this tomorrow.

**Comment by hugovk** [comment]
Thanks @lysnikolaou!

**Pull request merged**